Port over stdout implementation to wasmfs.cpp

ethanalee opened the pull request:
- Basic template for syscalls in wasmfs.cpp
- Build with `-s WASMFS` flag
Please see issue #15041 for background information regarding this PR

tlively commented on src/library_wasmfs.js:
```diff
@@ -0,0 +1,8 @@
+var WASMFS = {
```
Do you think we'll be filling out this implementation soon? If so, including this file might be worth it if you add a comment explaining that it is a placeholder for future work with a link to the tracking issue. Otherwise adding this file would probably be best left to when we have something to put in it.

tlively commented on emcc.py:
```diff
@@ -1767,6 +1767,12 @@ def default_setting(name, new_default):
+    state.forced_stdlibs.append('libwasmfs')
+    settings.FILESYSTEM = 0
+    settings.SYSCALLS_REQUIRE_FILESYSTEM = 0
+    # settings.JS_LIBRARIES.append((0, 'library_wasmfs.js'))
```
If we decide not to include `library_wasmfs.js` in this PR, then this could be a TODO comment mentioning that it should be added in the future instead.

tlively commented on src/settings.js:
```diff
@@ -1635,6 +1635,10 @@ var FETCH = 0;
 // [link]
 var ASMFS = 0;
 
+// If set to 1, uses new filesystem implementation
```
It would be good to add a note that this is still an experimental work in progress and is not expected to function correctly, probably with a link to the tracking issue.

ethanalee replied:
done

tlively commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+
+extern "C" {
+
+  __wasi_errno_t __wasi_fd_write( __wasi_fd_t fd, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
```
This line looks longer than 80 characters, so it would be nice to run `git clang-format` to standardize the formatting of the C/C++ source. @kripken or @sbc100, is that standard practice in this repo? Perhaps we should have CI check formatting like we do in Binaryen.

On the other hand, maybe clang-format does something terrible to `EM_ASM` macros :/

ethanalee replied:
I can use clang-format if that's appropriate for sure!

tlively commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+extern "C" {
+
+  __wasi_errno_t __wasi_fd_write( __wasi_fd_t fd, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
+    if (fd == 1 || fd == 2) {
```
This hardcoding of file descriptors 1 and 2 is temporary, right? It would be good to add a comment saying what this code does and that it is temporary behavior just to get something running at all.

juj replied:
I think it should be a permanent convention. We don't want to tie printing to require to pull in the filesystem. Making the filesystem somehow define stdio file handles would certainly create such a dependency.

kripken replied:
Yes, I agree that's an important goal here.

I'm hoping that we can find a way to make that happen even without hardcoding stuff here, just relying on LLVM. But if we don't, let's hardcode.

As background, currently we avoid linking in any FS code when only stdout/stderr are used here:

https://github.com/emscripten-core/emscripten/blob/aca7c128fd844d2b2b306aa2d5205a07f053bbe0/emscripten.py#L65

For WASMFS we could perhaps do something similar. I was hoping we'd find a less hackish mechanism, perhaps using weak linking, but I think all options should be on the table.

tlively commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+// // Copyright 2021 The Emscripten Authors.  All rights reserved.
+// // Emscripten is available under two separate licenses, the MIT license and the
+// // University of Illinois/NCSA Open Source License.  Both these licenses can be
+// // found in the LICENSE file.
```
It would probably be good to create `system/lib/wasmfs/wasmfs.cpp` rather than putting this under `system/lib/fetch`.

It would also be good to add a file-level comment saying what this file is for and what its implementation status is.

kripken commented on src/library_wasmfs.js:
```diff
@@ -0,0 +1,6 @@
+var WASMFS = {
+    $WASMFS : {}
+
```
```suggestion
  $WASMFS : {}
```

kripken commented on src/modules.js:
```diff
@@ -93,7 +96,9 @@ var LibraryManager = {
-    if (FILESYSTEM) {
+    if (WASMFS) {
+      libraries.push('library_wasmfs.js')
+    } else if (FILESYSTEM) {
```
As emcc.py sets `FILESYSTEM=0`, I don't think this is needed.

kripken commented on src/modules.js:
```diff
@@ -93,7 +96,9 @@ var LibraryManager = {
 
-    if (FILESYSTEM) {
+    if (WASMFS) {
+      libraries.push('library_wasmfs.js')
```
This could be moved up to be an else in that other if (if we keep it).

kripken commented on src/settings.js:
```diff
@@ -1635,6 +1635,10 @@ var FETCH = 0;
 // [link]
 var ASMFS = 0;
 
+// If set to 1, uses new filesystem implementation
```
Maybe add some notes here, like
* Experimental for now, try at your own risk.
* Intended to eventually replace most of the current JS filesystem code.

ethanalee replied:
This makes sense

tlively commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+        var nwritten = $3;
+        var num = 0;
+        for (var i = 0; i < iovs_len; i++) {
+          var ptr = HEAP32[(((iovs)+(i*8))>>2)];
+          var len = HEAP32[(((iovs)+(i*8 + 4))>>2)];
```
@kripken, is there a better way to write these memory accesses in an `EM_ASM`?

Stepping back, it would much easier to read if all this code that reads and writes memory were just normal C++ rather than JS. There shouldn't be any need to call out to JS for anything except the call to `out` or `err`, and `buffer` can probably be an ArrayBuffer initialized directly from a slice of the Wasm memory rather than built up one character at a time.

kripken replied:
This is basically the best way from JS, I think. (Though the extra parens can be removed, as I commented after you posted.)

We could do more work in C++ and leave less for JS. That would add additional copying, though, so I'm not sure it's a good idea.

tlively commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+  {
+    abort();
+  }
+}
```
Please add newlines at the end of your files!

kripken commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+#include <emscripten/emscripten.h>
+#include <wasi/api.h>
+
+extern "C" {
```
For C++ style, perhaps we can just use the clang-format settings for say LLVM or Binaryen? We can also leave that for later, but meanwhile, I'd suggest we use something more like the existing code we have, like `system/lib/embind/bind.cpp` (in particular: no indentation inside extern C)

curiousdannii replied:
Previously projects could just be compiled in C. Are there likely to be any implications of now adding in a C++ library?

ethanalee replied:
IIUC, @curiousdannii exported functions need to be C functions (to avoid C++ name mangling). If this is what you're asking.

ethanalee replied:
@kripken I used clang-format for now for consistency

sbc100 replied:
The main idea is that we don't indent stuff that lives inside `namespace {` or inside `extern "C" {`

ethanalee replied:
Got it. I think when I ran `clang-format` locally it removed the indentations.

tlively reviewed (commented):
Very cool to see this implementation starting up! It would be good to add a test (in test_core.py or maybe test_other.py for now? @kripken, WDYT?) at least to show that the library is built and linked in successfully, even if there isn't much functionality to test yet.

kripken commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+
+  __wasi_errno_t __wasi_fd_write( __wasi_fd_t fd, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
+    if (fd == 1 || fd == 2) {
+      MAIN_THREAD_EM_ASM({
```
Might be worth writing a TODO here so we don't forget for later: We may not want to proxy stderr to the main thread (it doesn't show up in the HTML anyhow - doing console.warn in a worker would be fine). But that would be a change from the current FS.

kripken commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+        var nwritten = $3;
+        var num = 0;
+        for (var i = 0; i < iovs_len; i++) {
+          var ptr = HEAP32[(((iovs)+(i*8))>>2)];
```
```suggestion
          var ptr = HEAP32[iovs + i*8 >> 2];
```
The JS order of operations allows this to be written without parens.

kripken commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+
+  __wasi_errno_t __wasi_fd_seek(__wasi_fd_t fd, __wasi_filedelta_t offset, __wasi_whence_t whence, __wasi_filesize_t *newoffset)
+  {
+    abort();
```
Might be nice to make a function that gets a name and prints an error out before aborting, so the runtime error is a little clearer. But it would help mostly you @ethanalee so up to you...

ethanalee replied:
I can have it print out the name of the stubbed function? @kripken

sbc100 replied:
I think you can: https://gcc.gnu.org/onlinedocs/gcc/Function-Names.html

ethanalee replied:
awesome!

kripken commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+
+extern "C" {
+
+  __wasi_errno_t __wasi_fd_write( __wasi_fd_t fd, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
```
```suggestion
  __wasi_errno_t __wasi_fd_write(__wasi_fd_t fd, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
```

kripken commented on src/modules.js:
```diff
@@ -61,14 +61,17 @@ var LibraryManager = {
+    if (!WASMFS) {
+      libraries.push('library_syscall.js')
+      libraries.push('library_wasi.js')
+    }
```
Given that we realized this files contain non-filesystem syscalls too, we might want to remove this if. Things should still work for you, although the error messages might be less clear - basically, any time you see `var FS` being included, that'd mean you have a syscall you need to implement. But, if the error messages are useful, then given this is still an experimental option that seems fine to me.

sbc100 replied:
I suggest instead you either:

1) modify these files and wrap the FS syscalls in `if !WASMFS`
2) split these out into `library_wasi_fs.js` and `'library_syscall_fs.js'

kripken reviewed (commented):
Great start!

For landing, let's add a test. Perhaps adding a variation on hello world,

https://github.com/emscripten-core/emscripten/blob/592dec1c71516c3c7535443e6d88ff5ad28db304/tests/test_core.py#L310-L314

One way to add a variation could be to add a decorator like that existing `@also_with_standalone_wasm` (so, `@also_with_wasmfs`). That might make sense, I think. And then we'd decorate more and more tests as things get working.

kripken commented:
Oh, and it would be good to link to #15041 from the relevant PRs. Something like `see #15041` in the top comment and the commit message perhaps.

kripken commented:
(sorry for racing with your feedback @tlively ! some overlap there...)

sbc100 commented on src/preamble.js:
```diff
@@ -617,7 +617,7 @@ function abort(what) {
 #include "memoryprofiler.js"
 
-#if ASSERTIONS && !('$FS' in addedLibraryItems) && !ASMFS
+#if ASSERTIONS && !('$FS' in addedLibraryItems) && !ASMFS && !WASMFS
```
Assuming the plan is to completely remove/replace ASMFS eventually can we remove ASMFS first to avoid having them in tree at the same time?

ethanalee replied:
May put this in another PR if needed?

sbc100 replied:
Yes, I just wonder if that PR should land first?  i.e. should we cleanup the old code before landing new stuff?

tlively replied:
We haven’t discussed removing asmjs with @juj yet, so keeping it in for now was deliberate.

sbc100 commented on src/library_wasmfs.js:
```diff
@@ -0,0 +1,6 @@
+var WASMFS = {
+    $WASMFS : {}
```
It could be that you don't need this global object to hang stuff off.. it has proved to be an anti-pattern in some cases because it can defeat DCE.

sbc100 commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+  __wasi_errno_t __wasi_fd_write( __wasi_fd_t fd, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
+    if (fd == 1 || fd == 2) {
+      MAIN_THREAD_EM_ASM({
+        var buffer = [];
```
Given that you have a JS library file I think it makes sense to put any large chunks of JS code there rather than inlining in C++.. for one thing it makes editing a lot easier (syntax highlighting, etc).

sbc100 commented on system/lib/fetch/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+// // Copyright 2021 The Emscripten Authors.  All rights reserved.
```
The issue that @curiousdannii brings up about using C++ in core libraries is worth thinking about.

The issues is that up until now one could build a C program and link agianst only C libraries and avoid the C++ runtime completely.  (i.e. use `emcc` rather then `em++`.. as of today we still make C++ libraries available unless `-s STRICT` mode is also used but I hope to change that).

Using C++ language features should be fine but if we use C++ library features in our filesystem implementation then it makes it much hard to build a pure C application. 

We do have some core libraries such as embind that are writting in C++ but those are not used to implement low level C APIs and its generally accepted that embind only works for C++ programs.

Its maybe not a huge issue, but if we do use C++ (and C++ library features in particular) that does means that users of WASMFS would need to make sure they link in libc++ and libc++abi.

kripken replied:
Interesting point. Does that mean that even using `new` would be an issue, as that method is in libc++abi I think? If so that is concerning.

I do think it would be a big productivity win to use as modern a language as we can here.

sbc100 replied:
Yes, I agree it nice to be able to implement system libraries in C++... and I think we should probably allow it here, but traditionally the C library implementation is implement in C .. for this reason.. otherwise it makes all program into C++ programs.

Really we are implementing part of the "OS" in C++ here but they happen to get linked into the user program.  

I guess we can just have `-s WASMFS` imply `-s DEFAULT_TO_CXX`.

kripken replied:
> I guess we can just have -s WASMFS imply -s DEFAULT_TO_CXX.

Yes. But note that the goal is to eventually enable WASMFS by default, and then even remove the option so it is the only thing we support, which would mean that any program using a nontrivial file-related syscall would end up a C++ program. So we need to make a decision here I guess.

sbc100 replied:
I guess it depends how much benefit we see from using C++ here.  Perhaps its not the end of the world if all filesystem-using programs end up depending on C++.

tlively replied:
Maybe we can bundle C++ runtime dependencies into the wasmfs build somehow so they don't need to be separately linked in?

sbc100 replied:
`DEFAULT_TO_CXX` should be sufficient to guarantee they are linked it.

curiousdannii replied:
Could LTO versions of libc++ etc be included in the Docker build? Prior to this you'd only have to wait for it to be built when your own code was linking against libc++, but now it will be all C projects that use the FS too.

sbc100 replied:
Yes, the Docker build should contains all flavors of libc++ i think.  IIUC its a bug that it doesn't.

kripken commented on tests/test_core.py:
```diff
@@ -313,6 +319,11 @@ def test_hello_world(self):
     # must not emit this unneeded internal thing
     self.assertNotContained('EMSCRIPTEN_GENERATED_FUNCTIONS', read_file('test_hello_world.js'))
 
+  @also_with_wasmfs
```
Sorry if I wasn't clear, I was suggesting adding a new decorator to the existing test. That would avoid adding a new test entirely, just one line for the decorator. (Well, and also defining the decorator the first time.)

ethanalee replied:
Oh apologies. I will delete the new test I added!

kripken commented on tests/test_core.py:
```diff
@@ -139,6 +139,12 @@ def can_do_standalone(self):
+def also_with_wasmfs(func):
+  def decorated(self):
+    self.set_setting('WASMFS')
+  return decorated
```
This doesn't have the behavior of "also" - it just tests with wasm FS. The `also_` family of decorators run the test once normally, and then once with an extra flag set, and I think that would make sense to do here too.

kripken commented on system/lib/wasmfs/wasmfs.cpp:
```diff
@@ -0,0 +1,72 @@
+  // TODO: May not want to proxy stderr (fd == 2) to the main thread
+  // This will not not show in HTML, a console.warn in a worker is suffficient
+  // This would be a change from the current FS
+  if (fd == 1 || fd == 2) {
```
Please add `.` at the end of sentences. (We're not 100% consistent there, but let's try in new code :smile: )

kripken commented on src/library_wasmfs.js:
```diff
@@ -0,0 +1,19 @@
+var WasmfsLibrary = {
+$WASMFS: {
```
Let's refactor this to avoid the single global `WASMFS` object. As @sbc100 mentioned that is a pattern that is not optimal because of code size.

Instead, something like

```js
$wasmfsBuffers: [null, [], []],

emscripten_wasmfs_printchar__deps: ['$wasmfsBuffers'],
emscripten_wasmfs_printchar: function(stream, curr) {
  // use wasmfsBuffers[1] etc.
```
That way we define plain C functions for the C++ to call, and those functions can add dependencies on other JS things they need.

In the C++ file we'd declare
```c
int emscripten_wasmfs_printchar(..);
```
and then call that - without knowing it is actually calling into JS.

Note that this also avoids the `EM_ASM` in the C++, which is also good.

kripken commented on system/lib/wasmfs/wasmfs.cpp:
```diff
@@ -0,0 +1,57 @@
+      const uint8_t* ptr = iovs[i].buf;
+      __wasi_size_t len = iovs[i].buf_len;
+      for (__wasi_size_t j = 0; j < len; j++) {
+        EM_ASM({ WASMFS.printChar($0, HEAPU8[$1]); }, fd, ptr + j);
```
It would be better to send an entire buffer to JS, and not a single character at a time, as calls to JS are not cheap.

In fact this may be an issue if we get many tiny buffers. That is something we'll need to measure at some point I guess - maybe add a TODO.

ethanalee replied:
Changed it to just use the buffers that we receive via `iovs`

kripken commented on system/lib/wasmfs/wasmfs.cpp:
```diff
@@ -0,0 +1,57 @@
+  __wasi_fd_t fd, __wasi_filedelta_t offset, __wasi_whence_t whence, __wasi_filesize_t* newoffset) {
+  EM_ASM({out($0 + ' or __wasi_fd_seek has been temporarily stubbed and is inert')},
+    __PRETTY_FUNCTION__);
+  abort();
```
Hmm, it is unfortunate to use `EM_ASM` here actually - it increases code size to add the JS integration. But we can't rely on C++ to print because this is the actual printing code :smile: Anyhow, as we will implement these, this lgtm.

ethanalee replied:
Mmm, yes there is certainly a circular dependency 😄

sbc100 replied:
We have the lower level `emscipten_console_log` which you could use here.  (we don't have a `console_log_printf` flavor bug we could add one).

https://github.com/emscripten-core/emscripten/blob/12b2b675c202d2abc0a3f6c756939dde39eb199e/system/include/emscripten/html5.h#L500`

kripken commented on system/lib/wasmfs/wasmfs.cpp:
```diff
@@ -0,0 +1,57 @@
+    for (size_t i = 0; i < iovs_len; i++) {
+      const uint8_t* ptr = iovs[i].buf;
+      __wasi_size_t len = iovs[i].buf_len;
+      emscripten_wasmfs_printchar(fd, ptr, len);
```
The name should be updated too, as it's no longer printing a single char. How about `_printbuffer`?

kripken commented:
The test errors look related to dynamic linking. I suspect that is due to recent work @sbc100 has been doing. Merging in very latest `main` should fix it, as then all the repos will be in sync (atm, this is using old emscripten code against newer LLVM and Binaryen, which we use the latest versions of on CI).

kripken commented on system/lib/wasmfs/wasmfs.cpp:
```diff
@@ -0,0 +1,55 @@
+// // See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include <stdlib.h>
+#define __NEED_struct_iovec
```
what does this do?

ethanalee replied:
Oh, I can remove that, it was from a prior definition of `iovec`. Realized that we already have it defined in `api.h`. Thanks

kripken reviewed (approved):
Great!

sbc100 commented on system/lib/wasmfs/wasmfs.cpp:
```diff
@@ -0,0 +1,54 @@
+// // found in the LICENSE file.
+// // wasmfs.cpp will implement a new file system that replaces the existing JS filesystem.
+// // Current Status: Work in Progress.
+// // See https://github.com/emscripten-core/emscripten/issues/15041.
```
Can you remove the double comments here?

sbc100 commented on tests/test_core.py:
```diff
@@ -139,6 +139,18 @@ def can_do_standalone(self):
 
+def also_with_wasmfs(func):
+  def decorated(self):
+    self.set_setting('WASMFS', 0)
```
I don't think you need this line.

ethanalee replied:
I was putting it there for clarity, but I guess it's redundant

ethanalee commented:
Please see issue #15041 for background information regarding this PR
- Added a basic template for syscalls in the new file system implementation. STDOUT is currently functional
- Added flag `-s WASMFS` to turn on this experimental feature
- Added @also_with_wasmfs decorator to existing tests

juj commented on src/library_wasmfs.js:
```diff
@@ -0,0 +1,17 @@
+    }
+    
+  }
+}
```
I would recommend implementing this whole function in C/C++. That way the state will be multithreading safe from the get-go (e.g. for behavior like `fflush()`ing stdout/stderr from another thread than the one(s) that have done partial printing). That method could use e.g. `emscripten_console_log()` and `emscripten_console_error()` for the printing.
